ERTP guide revision (#237)
Agoric/documentation: main/ertp/ERTP-Guide/Overview.md

```js
const { issuer, mint } = makeIssuerKit('BaytownBucks');
const bb1000 = amountMath.make(1000);
const newPayment = mint.mintPayment(bb1000);
const bbpurse = issuer.makeEmptyPurse();
const bb1000Amount = bbpurse.deposit(newPayment);
```

The act of minting created 1000 BaytownBucks and stored them together in a
`payment`. We then deposited the `payment` into a `purse` for BaytownBucks.

Now, let's distribute the BaytownBucks to others. To send assets
in ERTP, we withdraw a `payment` from a `purse`.

```js
const paymentForAlice = purse.withdraw(bucks10);
```

Like our `purse`, this `payment` contains BaytownBucks, but unlike purses,
we use payments to represent tokens in transit. A `payment` can be
sent to someone else, a `purse` should never be sent to someone else.

Now let's send the `payment` to Alice.

Before she can receive a `payment` in BaytownBucks, Alice needs to 
create an empty `purse` for BaytownBucks and otherwise get ready to
receive payments. 

For this to happen, Alice needs access to `baytownBucksIssuer()`.
However, she does not need access to `baytownBucksMint()`.
If she had access to the `mint()`, she could create BaytownBucks herself

> by calling `baytownBucksMint.mint()`. 

Reviewer: This also needs to be updated for current ERTP. I think just checking against the current ERTP API docs should work.
```suggestion
by calling `baytownBucksMint.mintPayment()`. 
```

```js
const aliceBaytownBucksPurse = baytownBucksIssuer.makeEmptyPurse()
aliceBaytownBucksPurse.getAmount(); // Empty, so the value is 0
```
Next, we make a deposit facet for Alice's `purse`. A deposit facet is an object 
associated with a `purse` which only allows deposits to its `purse` and cannot be used
to make a withdrawal from its `purse`.
```js
const aliceDepositFacet = purse.makeDepositFacet();
```
Here `aliceDepositFacet` is an object reference, and we can call its `receive` method
to deposit this `payment` into its associated `purse`. 
```js
aliceDepositFacet.receive(paymentForAlice);
```

And here is a more complete example using Agoric's Board, a simple "bulletin board"
for posting and getting Id-value pairs:
```js
// Alice posts an Id-value pair to the Board, where the value is the deposit facet object
// for her `purse`.
// Alice then gives us the boardId to use for the transfer: aliceDepositFacetBoardId

> // We retrieve the value from the boardID and then deposit the payment in the deposit facet.

Reviewer: thanks for adding this! I think it will be very helpful for our developers in our current system

# ERTP Guide

[ERTP](https://agoric.com/documentation/glossary/#ertp) (*Electronic
Rights Transfer Protocol*)
is Agoric's token standard for transferring tokens and other digital assets in
JavaScript. Using the
[ERTP API](https://agoric.com/documentation/ertp/api/),
you can easily create and use digital assets, all of which are
transferred exactly the same way and with exactly the same security properties. 

ERTP uses
*[object capabilities](https://agoric.com/documentation/glossary/#object-capabilities)*
to enforce access control. If your program has a reference to an
object, it can call methods on that object. If it doesn't have a
reference, it can't. For more on object capabilities, see
[Chip Morningstar's post](http://habitatchronicles.com/2017/05/what-are-capabilities/).

## Objects Overview

> ![ERTP object relationships](./assets/relationships.svg)  

Reviewer: I think this diagram will be more confusing than helpful, but the idea of having a diagram is very useful. I think this particular diagram doesn't mention digital assets and doesn't have enough of a hierarchy - it seems like everything is at the same level. Can we switch to something like this:

![IMG_20200731_151022](https://user-images.githubusercontent.com/2441069/89081753-e285ab00-d340-11ea-8b15-f9af698d1e35.jpg)

Brand shouldn't be its own section in the diagram, but should be describing everything else. Digital assets should be in the center. Amounts should be off of digital assets. We may want to explain more of what the issuer does in detail in a separate diagram. I don't have very good ideas for that, but here's a representation of what is stored in the issuer.
![IMG_20200731_150858](https://user-images.githubusercontent.com/2441069/89081839-21b3fc00-d341-11ea-9419-87415779cb80.jpg)

> There are eight fundamental ERTP objects, two of which are parts of

Reviewer: Mentioning the exact number and which are parts of another does not seem to be helpful for the reader. The reader wants to know 1) what ERTP is and 2) how they can use it quickly. This is not either of those.

Reviewer: Also, I think we mention objects a lot. Why? Why say `payment object` when we could just say `payment`?

Author: Honestly, a fair number of those were to be able to put the "s" on "objects" rather than `purse`s, which looks odd.

Reviewer: ah, ok! I think `purses` looks fine if we need to make it plural.

another, and one which is used by that same object. For each entry,
the name is linked to its primary page in this Guide. 

> - **[Amount](./Amounts.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#amount)**:

Reviewer: Let's not start with Amounts, let's start with Digital Assets. Then explain that Mints are the only thing that mints them. And that they are stored in Purses and Payments. 

To explain Issuers, I think we should start from the scenario of receiving a payment from someone you don't know. An Issuer is the source of truth for whether it is a valid payment.

>   The description of an asset, it is not an asset itself, having no economic scarcity or economic value.

Reviewer: run on

  Made up of a:
  - *[value](./Amounts.md)* *[(Glossary)](https://agoric.com/documentation/glossary/#value)*:  The
  size of the amount.
  - *[brand](./Amounts.md)* *[(Glossary)](https://agoric.com/documentation/glossary/#brand)* The
  kind of asset. 
  
  You can think of these as the "how many" and the "what" of an asset, for example, 7 quatloos (an imaginary
  currency).
- **[AmountMath](./AmountMath.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#amountmath)**:
Methods to do math operations on an `amount`..

> - **[Purse](./PursesAndPayments.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#purse)**: an

Reviewer: Good point. I agree

  object for long term digital assets storage. 

> - **[Payment](./PuresAndPayments.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#payment)**:

Reviewer: ```suggestion
- **[Payment](./PursesAndPayments.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#payment)**:
```

  an object for transfering digital assets to another party.
- **[Mint](./IssuersAndMints.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#mint)**: an

>   object that creates new `payment` objects of a specific `brand`. Each `brand` has

Reviewer: ```suggestion
  the only way to mint digital assets of a particular brand. The assets are stored in payments.
```

  a one to one relationship with a `mint` and vice versa.
- **[Issuer](./IssuersAndMints.md)** **[(Glossary)](https://agoric.com/documentation/glossary/#issuer)**: 

> an object linked to a single `mint` object that creates empty `purse` objects

Reviewer: I think this doesn't define the most important thing about issuers, and mentions objects too much. 

How about: "the source of truth of how many digital assets each purse and payment holds. The issuer can be used to validate payments received by untrusted parties."

for that mint's `brand` and operates on any `payment` of that `brand`. 

## Fungible and Non-Fungible Assets

There are two kinds of assets,

> *[fungible](https://agoric.com/documentation/glossary/#fungible)* and

Author: I think I'm going to leave this one. Working backwards, "non-fungible" gives an easy to use understandable term to use for them. I used "kind" for the nat, set, strSet to avoid overloading "type" and confusing it with JS typing and the like. I think a quick use of it in plain English  here, particularly since we never use "fungible" in ERTP itself, is OK.

*[non-fungible](https://agoric.com/documentation/glossary/#non-fungible)*. 

Fungible assets are interchangeable. For example, if you have 100
one-dollar bills and need to pay someone 5 dollars, it doesn't matter
which five of your one-dollar bills you give them. 

Non-fungible assets are of the same type, but are not interchangeable
and specific items must be used. For example, you might have 100
tickets. But someone wanting to buy even a General Admission ticket from you will want one
for a specific date and time. This might also affect the price; you'll want to charge more
for a Friday evening ticket than a Wednesday matinee ticket, even if it's for the same show.

## Amounts are not assets

**IMPORTANT**: Despite how it may seem, an `amount` is only a description of an asset, not

> an asset in and of itself. Actual assets are in `purse` and `payment` objects, but the

Reviewer: ```suggestion
an asset in and of itself. Actual assets are in `purses` and `payments`, but the
```

assets aren't ever embodied in any particular JavaScript object. Instead, what makes the 

> assets exist is the issuer having a mapping from the `purse` or `payment` to an `amount` (description).

Author: I'd like to get Kate's opinion on this. She and I had some back and forth on how I had the mistaken belief that there were amount objects effectively stored in the p/p objects, and I think your phrasing is more likely to cause people to get that belief than an explicit "No, that's not how it works".

> assets exist is the issuer having a mapping from the `purse` or `payment` to an `amount` (description).

Reviewer: How about "assets are held by purses and payments in the same way that a bank account holds money - it does so not because of anything physical, but because of a record held by an authoritative source, in this case, the issuer is the source of truth of what assets are held by what purses and payments"

There is no `amount` object stored in a `purse` or `payment`.

Some analogies may be helpful in understanding this:

> - Let's say Caltech replicates its [famous hack of the Rose Bowl scoreboard](https://www.admissions.caltech.edu/pranks).

Reviewer: I think these examples do not get at the right point - changing the amount balance in the records held by the issuer would actually change the digital assets. It's whether the record is the source of truth or not that actually matters.

  actually used in contracts**

## Object capabilities and ERTP

ERTP implements [*object capabilities*](https://agoric.com/documentation/glossary/#object-capabilities). You can only use an object and issue commands to it if you have access to that object, not just its human-readable name or similar. For example, I might know (or make a good guess), that the mint that makes quatloos has the human-understandable alleged name of 'quatloos-mint'. But unless I have the actual object that is a `mint` object associated with the `quatloos` `brand` object, I can't use it to create a million `quatloos` and bet
them all on Captain Kirk to win his gladiatorial match on Triskelion (see the [Wikipedia entry for the Star Trek episode](https://en.wikipedia.org/wiki/The_Gamesters_of_Triskelion)).

## Creating and sending a fungible asset

In ERTP, digital assets are created by a `mint`. You must have
access to a `mint` object to create new assets of the `mint` object's
associated `brand`. 

For instance, let's say we want to create a new community
currency called 'BaytownBucks'. You start by getting access to a `mint`

> object whose associated brand is 'BaytownBucks'.

Author: Changed to "by creating or getting access to"

```js

> import { makeMint } from '@agoric/ertp';

Reviewer: This needs to be updated for current ERTP - it's makeIssuerKit

const baytownBucksMint = makeMint('BaytownBucks');
```

Now, use the `mint` to create 1000 new BaytownBucks.

```js
const { issuer, mint } = makeIssuerKit('BaytownBucks');
const bb1000 = amountMath.make(1000);
const newPayment = mint.mintPayment(bb1000);
const bbpurse = issuer.makeEmptyPurse();
const bb1000Amount = bbpurse.deposit(newPayment);
```

The act of minting created 1000 BaytownBucks and stored them together in a
`payment`. We then deposited the `payment` into a `purse` for BaytownBucks.

Now, let's distribute the BaytownBucks to others. To send assets
in ERTP, we withdraw a `payment` from a `purse`.

```js
const paymentForAlice = purse.withdraw(bucks10);
```

Like our `purse`, this `payment` contains BaytownBucks, but unlike purses,
payments are used to represent tokens in transit. A `payment` can be

> sent to someone else, a `purse` should never be sent to someone else.

Author: I think the current phrasing works with that  I did edit it a bit to make it active, but "unlike purses, we use payments to represent tokens in transit" doesn't per se forbid purses to be used for it, just says how we use them. Similarly "a purse should never be sent" has that nice weasel word "should" so we're not forbidding it.

# ERTP Guide

[ERTP](https://agoric.com/documentation/glossary/#ertp) (*Electronic
Rights Transfer Protocol*)
is Agoric's token standard for transferring tokens and other digital assets in
JavaScript. Using the
[ERTP API](https://agoric.com/documentation/ertp/api/),
you can easily create and use digital assets, all of which are
transferred exactly the same way and with exactly the same security properties. 

ERTP uses
*[object capabilities](https://agoric.com/documentation/glossary/#object-capabilities)*
to enforce access control. If your program has a reference to an
object, it can call methods on that object. If it doesn't have a
reference, it can't. For more on object capabilities, see
[Chip Morningstar's post](http://habitatchronicles.com/2017/05/what-are-capabilities/).

## Objects Overview

> ![ERTP object relationships](relationships.svg)  

Reviewer: I think this diagram is very confusing and parts are not quite right. The mint doesn't create amounts. AmountMath should be in here alongside brand and issuer and mint. purses and payments contain digital assets, not amounts per se. Amount doesn't use amountMath.

> ![ERTP object relationships](relationships.svg)  

Reviewer: Let's work on this together.

There are eight fundamental ERTP objects, two of which are parts of
another, and one which is used by that same object. For each entry,
the name is linked to its section in this document. The following "&"

> symbol is linked to the term's [Glossary]() entry.

Reviewer: I think the & symbol might be better replaced with`Glossary`

- **[Amount](#amount)** **[&](https://agoric.com/documentation/glossary/#amount)**:

>   the representation of an asset, made up of a

Reviewer: It's really important that we call this a description - it is not the asset itself. It has no economic scarcity or economic value.

```suggestion
  the description of an asset, made up of a
```

  *[value](#value)* *[&](https://agoric.com/documentation/glossary/#value)*  (the
  size of the amount) and a
  *[brand](#brand)* *[&](https://agoric.com/documentation/glossary/#brand)* (the
  kind of asset). You can think of these as the "how many" and the
  "what" of an asset.

> - **[AmountMath](#amountmath)** **[&]((https://agoric.com/documentation/glossary/#amountmath)**:

Reviewer: parens are off

> Methods to do math operations on `amount` objects.

Reviewer: we should really remove any mention of `amount` objects and just say `amounts`

- **[Purse](#purse)** **[&](https://agoric.com/documentation/glossary/#purse)**: an

>   object for long term `amount` object storage. 

Reviewer: Assets are stored in purses, not amounts. Amounts are merely the descriptions of the assets stored in the purse.

Author: How about "for long term asset storage, where an `amount` describes the asset". ?

Reviewer: Why mention amounts?

- **[Payment](#payment)** **[&](https://agoric.com/documentation/glossary/#payment)**:

>   an object for transfering an `amount` object to another party.

Reviewer: No, this is very wrong. A payment is a transfer of economic value. The amount of digital assets transferred is the `amount`. But it is the assets that are transferred, not the amount. The amount is merely the description.

- **[Mint](#mint)** **[&](https://agoric.com/documentation/glossary/#mint)**: an

>   object that creates new `amount` objects of a specific brand. Each brand has

Reviewer: Mints mint payments. Not amounts.

  a one to one relationship with a mint and vice versa.
- **[Issuer](#issuer)** **[&](https://agoric.com/documentation/glossary/#issuer)**: 
an object linked to a single `mint` object that creates empty `purse` and
`payment` objects for that mint's brand and operates on payments of
that brand. 

## Fungible and Non-Fungible Assets

There are two kinds of assets,
*[fungible](https://agoric.com/documentation/glossary/#fungible)* and
*[non-fungible](https://agoric.com/documentation/glossary/#non-fungible)*. 

Fungible assets are interchangeable. For example, if you have 100
one-dollar bills and need to pay someone 5 dollars, it doesn't matter
which five of your one-dollar bills you give them. 

Non-fungible assets are of the same type, but are not interchangeable
and specific items must be used. For example, you might have 100
tickets. But someone wanting to buy a ticket from you will want one
for a specific date, a specific theater, a specific time, and let's
say a specific seat (these aren't General Admission tickets). 

> You can't just pull out a ticket at random and sell it to them, since

Reviewer: I think the reader won't agree with this as written. I think we mean that tickets between shows are not fungible, but tickets within shows might be for the user. We should explain this in a different way so it's more easily agreed with.

each ticket will have at least one different value from all the
others. This might also affect the price; you'll want to charge more
for a Friday evening third row center ticket than a Wednesday matinee
balcony ticket, even if it's for the same act.

## Security properties

ERTP `purse` objects have a `deposit` method which takes a `payment`
object as its argument. It first checks that the `payment` object is

> genuine and the same asset type as the `purse` (after all, our quatloos currency is
> a very unstable asset, so we wouldn't want to let people deposit
> quatloos and get credit for simoleans). Note: quatloos and simoleans
> are both imaginary currencies. 

Reviewer: ```suggestion
genuine and the same asset type as the `purse`. Note: quatloos and simoleans
are both imaginary currencies. 
```
I think the mention of it being unstable (where did that come from?) and simoleans is misleading and unnecessary. We could just say that one purse contains one type of asset.

If everything passes the checks, the asset moves from 
the `payment` to the `purse`. If there's a problem, it throws an error.

After a successful deposit, ERTP guarantees:

> - The `payment` object is empty.

Reviewer: the payment object is deleted from the issuer's records

Reviewer: This means that the issuer no longer recognizes it, and no longer has any amount associated with it.

JavaScript implements `Promise` objects, and recently added the two commands `async` and `await` to simplify working with them. For general, and extensive, information about JavaScript's implementation, see either:
- [javascript.info](https://javascript.info/async)
- [Mozilla's Developer Docs](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous)

**tyg todo: Needs beefing up about how Promises, async, and await are
  actually used in contracts**

## Alleged Names

ERTP implements [*object capabilities*](https://agoric.com/documentation/glossary/#object-capabilities). You can only use an object and issue commands to it if you have access to that object, not just its human-readable name or similar. For example, I might know (or make a good guess), that the mint that makes quatloos has the human-understandable name of 'quatloos-mint'. But unless I have the actual object that is a `mint` object associated with the `quatloos` `brand` object, I can't use it to create a million `quatloos` and bet
them all on Captain Kirk to win his gladitorial match on Triskelion (see the[ Wikipedia entry](https://en.wikipedia.org/wiki/The_Gamesters_of_Triskelion)).

> **tyg todo: Needs beefing up on how allegedNames are used in contracts

Reviewer: AllegedNames aren't used really at all, for the reason that they aren't trusted. If you have an object and you're not sure what it was intended for, you might peek at the name though, and then make other calls that are trustworthy.

  vs. getting and using the real objects**

## Creating and sending a fungible asset

> In ERTP, digital assets are created by a [`mint`](). You must have

Reviewer: link?

access to a `mint` object to create new assets of the `mint` object's
associated `brand`. 

For instance, let's say we want to create a new community
currency called 'BaytownBucks'. You start by getting access to a `mint`
object whose associated brand is 'BaytownBucks'.
```js
import { makeMint } from '@agoric/ertp';
const baytownBucksMint = makeMint('BaytownBucks');
```

Now, use the `mint` to create 1000 new BaytownBucks.

```js
const { issuer, mint } = produceIssuer('BaytownBucks');

> const BB1000 = amountMath.make(1000);

Reviewer: We shouldn't use variable names with starting capitals in JavaScript. That is usually reserved for classes

const newPayment = mint.mintPayment(BB1000);
const BBpurse = issuer.makeEmptyPurse();

> const BB100Amount = BBpurse.deposit(newPayment);

Reviewer: deposit returns the value of the payment, which is BB1000

```

The act of minting created 1000 BaytownBucks and stored them together in a
`payment`. We then deposited the `payment` into a `purse` for BaytownBucks.

Now, let's distribute the BaytownBucks to others. To send assets
in ERTP, we withdraw a `payment` from a `purse`.

```js

> const paymentForAlice = purse.withdraw(10, `alice's community money`);

Reviewer: This is using a very old API. We need to fix this.

```suggestion
const paymentForAlice = purse.withdraw(bucks10);
```

```

Like our `purse`, this `payment` contains BaytownBucks, but unlike purses,
payments are used to represent tokens in transit. A `payment` can be
sent to someone else, a `purse` should never be sent to someone else.

Now let's send the `payment` to Alice as a message:

```js

> alice.receivePayment(paymentForAlice);

Reviewer: We can actually do this now! I think we should explain how to use the board here and actually send a payment to another user. We might want to do this in a link that can be clicked through to.
Without mention of the board:

```suggestion
aliceDepositFacet.receive(paymentForAlice);
```
With mention of the board: 
```js
// Alice gives us the boardId to use for the transfer: aliceDepositFacetBoardId
const aliceDepositFacet = await E(board).getValue(aliceDepositFacetBoardId);
E(aliceDepositFacet).receive(payment);
// Alice got her payment automatically deposited in her purse!
```

```

Here `alice` is an object reference, and we can call her `receivePayment`method
to ask her to receive this payment. Alice's methods are entirely up to her, 
and are not part of ERTP.

### Receiving the asset

Before she can receive a `payment` in BaytownBucks, Alice needs to 
create an empty `purse` for BaytownBucks and otherwise get ready to
receive payments. 
For this to happen, Alice needs access to `baytownBucksIssuer`.
However, she does not need access to `baytownBucksMint`.
If she had access to the `mint`, she could create BaytownBucks herself
by calling `baytownBucksMint.mint`. 

```js
const myBaytownBucksPurse = baytownBucksIssuer.makeEmptyPurse()

myBaytownBucksPurse.getAmount(); // Empty, so the value is 0

const alice = {

>     receivePayment(allegedBaytownBucksPayment){

Reviewer: see comment above about the depositFacet and board to show how it should actually be done.

        myBaytownBucksPurse.depositAll(allegedBaytownBucksPayment)
    }
}
```

Now, when `alice.receivePayment(paymentForAlice)` from earlier 
is called, Alice receives the `payment` and deposits its 1000
BaytownBucks in her purse. 

When Alice wants to exchange something for 300 BaytownBucks, she
creates a `payment` from her `purse`: 
```js
const payment = myBaytownBucksPurse.withdraw(baytownBucks(300));
```
This came naturally without having to express "which" 300 she wanted to withdraw,
since BaytownBucks are fungible.

## Creating a non-fungible asset with ERTP

### Modeling and creating the asset

In ERTP, digital assets are created by a [`mint`](./mint.html). Having
access to the `mint` gives you the power to create more digital assets
of the same brand at  will.

Let's say we own an Opera and want to sell tickets to seats for ballet shows. Tickets are
the non-fungible assets we want to represent. Tickets refer to a
specific seat for a specific show at a specific time and date. 

**tyg todo: This has changed sufficiently from what's in the Guide
that I'd appreciate someone who really knows what they're doing
updating it**

> ```js
> import { noCustomization } from '@agoric/ertp/core/config/noCustomization.js';
> import { makeCoreMintKeeper } from '@agoric/ertp/core/config/coreMintKeeper';
> 
> import { insist } from '@agoric/insist';
> import { mustBeComparable } from '@agoric/same-structure';
> 
> const insistOptDescription = optDescription => {
>   insist(!!optDescription)`optDescription must be truthy ${optDescription}`;
>   mustBeComparable(optDescription);
> };
> 
> function makeBalletTicketConfig() {
>   return harden({
>     ...noCustomization,
>     makeMintKeeper: makeCoreMintKeeper,
>     valueOpsName: 'uniValueOps',
>     valueOpsArgs: [insistOptDescription],
>   });
> }
> 
> const balletTicketMint = makeMint('Agoric Ballet Opera tickets', makeBalletTicketConfig);
> ```
> 

Reviewer: ```suggestion
```js
import { makeIssuerKit } from '@agoric/ertp';
const { mint: balletTicketMint, amountMath: balletTicketAmountMath } = makeIssuerKit('Agoric Ballet Opera tickets', 'set');
```

At this Opera, there are [1114](https://fr.wikipedia.org/wiki/Grand_Th%C3%A9%C3%A2tre_(Bordeaux)#Salle_de_spectacle) seats numbered `1` to `1114`.
Objects that represent valid tickets will have the following properties:
- `seat` with a number
- `show` with a string describing the show
- `start` with a string representing a [time/date in ISO format](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString)

The first step in creating tickets is to create JavaScript objects,
each representing a ticket. Then, only
[units](https://agoric.com/documentation/ertp/guide/units.html) can be
minted, so let's create units from the JavaScript objects and then,
let's mint the tickets! 

```js
const startDateString = (new Date(2019, 11, 9, 20, 30)).toISOString();

> const ticketObjects = Array(1114).fill().map((_, i) => ({

Reviewer: ```suggestion
const ticketValues = Array(1114).fill().map((_, i) => ({
```

  seat: i+1,
  show: 'The Sofa',
  start: startDateString,
}))

> const ticketUnits = ticketObjects.map(ticketValue => balletTicketMint.getAssay().makeUnits(ticketValue));

Reviewer: ```suggestion
const ticketAmounts = ticketValues.map(ticketValue => balletTicketAmountMath.make(ticketValue));
```

> const balletTicketPurses = ticketUnits.map(ticketUnit => balletTicketMint.mint(ticketUnit))
> ```

Reviewer: ```suggestion
const balletTicketPayments = ticketAmounts.map(ticketAmount => balletTicketMint.mintPayment(ticketAmount))
```
```

> For each ticket unit, we've created a

Reviewer: ```suggestion
For each ticket amount, we've created a
```

> [purse](https://agoric.com/documentation/ertp/api/purse.html) which

Reviewer: ```suggestion
payment which
```

> contains the corresponding unit. These purses can later be transformed
> into
> [payments](https://agoric.com/documentation/ertp/api/payment.html) via
> a call to `balletTicketPurse.withdrawAll()`. This payment can then be
> bought and sold and used in smart contracts. 
> 

Reviewer: ```suggestion
contains digital assets of the corresponding amount. The digital assets in these payments can be transferred to other people and used in smart contracts.
```

## More ERTP resources

Agoric's Mark Miller explained ERTP on Oct 10, 2018 in his [Programming Secure Smart Contracts][watch] presentation
during San Francisco Blockchain Week at a
[SF Cryptocurrency Devs meetup](https://www.meetup.com/SF-Cryptocurrency-Devs/events/253457222/).

[![miller-sfbw-erights](https://user-images.githubusercontent.com/150986/59150095-b8a65200-89e3-11e9-9b5d-43a9be8a3c90.png)][watch]

> [watch]: https://www.youtube.com/watch?v=YXUqfgdDbr8

Reviewer: I think we can remove this, as it is 2 years old

  An object for transfering digital assets of a specific `brand` to another party.
  
Similar to other component instances, a `purse` and a `payment` only work with one
`brand`. So a `purse` or `payment` that holds Quatloos cannot hold an asset of `brand` Moola or vice versa. 
You cannot change the `brand` a `purse` or `payment` was originally associated with. Once you create a
"Quatloos `purse`" or "Quatloos `payment`", they can never hold anything other than Quatloos.

However, unlike the other components, these are not one-to-one relationships. There can be thousands or more
`purses` or `payments` that hold Quatloos or any other `brand`.

## Life of Assets

Let's look at some asset "lifecycles". While it's very rare for an asset to be destroyed, these "lifecycles"
show assets from their creation through common usage patterns. These are deliberately stripped down to their
basic, core, functionality. Optional parameters and non-core operations are not shown, nor are some significant
concepts which would make this introduction more confusing. These are covered on the component-specific pages.

### Asset creation and storage

![Asset creation](./assets/assetCreation.svg)
```js

> const { quatloosIssuer, quatloosMint, quatloosAmountMath, quatloosBrand } = makeIssuerKit('quatloos');

Reviewer 2: ```const { issuer: quatloosIssuer, mint: quatloosMint, amountMath: quatloosAmountMath, brand:  quatloosBrand } = makeIssuerKit('quatloos');```

> const { quatloosIssuer, quatloosMint, quatloosAmountMath, quatloosBrand } = makeIssuerKit('quatloos');

Reviewer 3: This is just wrong.  It needs to be either:
```js
const { issuer, mint, amountMath, brand } = makeIssuerKit('quatloos');
```
or
```js
const { issuer: quatloosIssuer, mint: quatloosMint, amountMath: quatloosAmountMath, brand: quatloosBrand } = makeIssuerKit('quatloos');
```

Reviewer 3: ```suggestion
const { issuer: quatloosIssuer, mint: quatloosMint, amountMath: quatloosAmountMath, brand: quatloosBrand } = makeIssuerKit('quatloos');
```

is @Chris-Hibbert's suggestion.

```
First, you take a string naming a new to the system `brand` and use it as the argument to
`makeIssuerKit()`. The name underestimates its power; in addition to returning a new `issuer`
for the specified `brand`, it also returns a new `mint`, `amountMath`, and formal `brand` 

> for the argument. All are in one-to-one associations with each other. 

Author: Good idea about the naming scheme, but I don't want to bury it in an otherwise unrelated section (or have it take attention away from what the section's actually about). I moved that to its own section just before this one, and will include it in an appropriate place in Zoe docs.

previous day at 15 dollars per share rather than the correct 1507 dollars. This does not cause Google stock to be valued at
15; it's not the actual asset, but a description of its value (again, in this case, an incorrect one).

Making a new `amount` or changing an `amount` `value` does not create any new assets. New assets are created by using 
a `mint` to create a new `payment` that contains a newly created asset. In the real world, I can't create a new ten dollar
bill by drawing one; a new one has to be minted (printed) by an authorized government-run facility with its asset status
derived from its government backing. 

So an `amount` just describes an asset along the two axes of how many and
what units it's in (`value` and `brand`). They're used as a way of negotiating
with parties that doesn't involve sending/sharing the actual asset
until a deal is made. In other words, I don't make you an offer that I'll swap you a ticket to *Hamilton* for something
by sending you an actual ticket. Instead I make you an offer by sending you a written description
of what I'm willing to swap, namely a *Hamilton* ticket. If the offer is accepted, then I send you the actual asset, 

> in this case an actual *Hamilton* ticket (enjoy the show!).

Author: I think I'd like to leave it generic. While we certainly expect them to be looking at Zoe as well, bringing in Zoe, deposits, and contracts just complicates things. The point here is there's a swap with offers happening, and giving someone a written "I'll swap a Hamilton ticket for $300"  isn't the same as either an actual ticket or an actual $300. I expanded it a bit to include $300 as the other side of the swap.
